### ISSUE TITLE
security.osclientcerts.autoload in Policy 2.3 is not working with ESR 78.3.0

### ISSUE THREAD

**Issue by olphger** [mozilla/policy-templates#688]
Hi,

i had activated the preference security.osclientcerts.autoload to 1 (True) since some ESR Versions.

Now i updated to ESR 78.3.0 and if i go to about:config i can see that die setting is not "true" anymore. its false now and its not blocked to change.

It should be true and blocked to change, because i set the GPO security.osclientcerts.autoload to True

ESR 78.2.0 is still working.

So i updated the Policy Templates from 2.2 to 2.3 but its the same.

**Comment by mkaply** [comment]
Unfortunately we discovered a bug that is causing legacy preference not to work on ESR 78.3.0 and Firefox 81. It's unrelated to the GPO.

The only workaround right now is to switch to the new JSON preference policy.

https://github.com/mozilla/policy-templates/blob/master/README.md#preferences

I'm working to get a fix in ASAP.

**Comment by Mgeige1** [comment]
Hi,

I think I'm experiencing the same issue. I'm pushing a profile like this and and boolean value is not honored on firefox

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
   <dict>
      <key>EnterprisePoliciesEnabled</key>
      <true />
      <key>PayloadDisplayName</key>
      <string>Firefox</string>
      <key>PayloadIdentifier</key>
      <string>org.mozilla.firefox.6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadType</key>
      <string>org.mozilla.firefox</string>
      <key>PayloadUUID</key>
      <string>6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadVersion</key>
      <integer>1</integer>
      <key>Preferences</key>
      <dict>
         <key>security.default_personal_cert</key>
         <string>Select Automatically</string>
         <key>security.osclientcerts.autoload</key>
         <true />
      </dict>
   </dict>
</plist>

**Comment by mkaply** [comment]
Yes, that's going to be the same problem because it's a boolean using the old preference API.

**Comment by Mgeige1** [comment]
thanks I wasn't clear on the workaround.  Im pushing


      <key>security.osclientcerts.autoload</key>
         <true />

**Comment by mkaply** [comment]
Wait, you're on Mac. This definitely shouldn't be happening on macOS. The specific problem was converting 0/1 in GPO to true/false.

What are you seeing in about:config?

**Comment by Mgeige1** [comment]
On about:config it shows as false.  Any other value I set that is a string gets configured and locked

Matt Geiger
R&D Management

On Sep 25, 2020 11:55 AM, Michael Kaply <notifications@github.com> wrote:

Wait, you're on Mac. This definitely shouldn't be happening on macOS. The specific problem was converting 0/1 in GPO to true/false.

What are you seeing in about:config?

**Comment by mkaply** [comment]
That's really odd. I didn't expect this to affect macOS. Easiest workaround is to use the new format for preferences. See the example here:

https://github.com/mozilla/policy-templates/blob/master/README.md#preferences
```
<dict>
  <key>Preferences</key>
  <dict>
    <key>security.osclientcerts.autoload</key>
    <dict>
      <key>Value</key>
      <true/>
      <key>Status</key>
      <string>locked</string>
    </dict>
  </dict>
</dict>
```

I'll run some tests to make sure this is fixed by my new patch.

**Comment by Mgeige1** [comment]
Hi Michael,

I tried this code and it will lock the preference but it is still stuck on false

Matt Geiger
R&D Management

**Comment by Mgeige1** [comment]
Even when I push that code, It locks it as false

Matt Geiger
R&D Management

**Comment by mkaply** [comment]
I clearly screwed up here with a misunderstanding of how plist boolean gets converted to boolean. Working on a fix.

**Comment by mkaply** [comment]
Here's an incantation that will work (I've tested). Working on fixing this in Firefox:
```
<key>Preferences</key>
  <string>{"security.osclientcerts.autoload": {"Value": true,"Status": "locked"}}</string>
```

**Comment by mkaply** [comment]
Even simpler:
```
<key>Preferences</key>
  <string>{"security.osclientcerts.autoload": true}</string>
```

**Comment by Mgeige1** [comment]
Thank you I will test it out

Matt Geiger
R&D Management

On Sep 25, 2020 7:40 PM, Michael Kaply <notifications@github.com> wrote:

Even simpler:

<key>Preferences</key>
  <string>{"security.osclientcerts.autoload": true}</string>

**Comment by Mgeige1** [comment]
Hi Michael,

This does work by itself, but if I add another preference with a string value then it will not work.  I listed the combinations below that I tried.

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
   <dict>
      <key>EnterprisePoliciesEnabled</key>
      <true />
      <key>PayloadDisplayName</key>
      <string>Firefox</string>
      <key>PayloadIdentifier</key>
      <string>org.mozilla.firefox.6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadType</key>
      <string>org.mozilla.firefox</string>
      <key>PayloadUUID</key>
      <string>6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadVersion</key>
      <integer>1</integer>
      <key>Preferences</key>
      <dict>
         <key>security.default_personal_cert</key>
         <string>Select Automatically</string>
           <key>Preferences</key>
        <string>{"security.osclientcerts.autoload": {"Value": true,"Status": "locked"}}</string>

      </dict>
   </dict>
</plist>

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
   <dict>
      <key>EnterprisePoliciesEnabled</key>
      <true />
      <key>PayloadDisplayName</key>
      <string>Firefox</string>
      <key>PayloadIdentifier</key>
      <string>org.mozilla.firefox.6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadType</key>
      <string>org.mozilla.firefox</string>
      <key>PayloadUUID</key>
      <string>6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadVersion</key>
      <integer>1</integer>
      <key>Preferences</key>
      <dict>
         <key>security.default_personal_cert</key>
         <string>Select Automatically</string>
           <key>security.osclientcerts.autoload</key>
        <string>{"security.osclientcerts.autoload": {"Value": true,"Status": "locked"}}</string>

      </dict>
   </dict>
</plist>

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
   <dict>
      <key>EnterprisePoliciesEnabled</key>
      <true />
      <key>PayloadDisplayName</key>
      <string>Firefox</string>
      <key>PayloadIdentifier</key>
      <string>org.mozilla.firefox.6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadType</key>
      <string>org.mozilla.firefox</string>
      <key>PayloadUUID</key>
      <string>6815559E-0936-40B8-8745-4028EFA1B667</string>
      <key>PayloadVersion</key>
      <integer>1</integer>
      <key>Preferences</key>
  <string>{"security.osclientcerts.autoload": {"Value": true,"Status": "locked"}}</string>

      <dict>

         <key>security.default_personal_cert</key>
         <string>Select Automatically</string>

      </dict>
   </dict>
</plist>
Matt Geiger
R&D Management

**Comment by Mgeige1** [comment]
Hi Michael,

I'm having trouble getting two preferences to apply in the same string.  I'm pushing the below code block.

      <key>Preferences</key>
 
      <string>{"security.default_personal_cert": "Select Automatically",
"security.osclientcerts.autoload": true
</string>

**Comment by mkaply** [comment]
Missing a closing bracket.

try this:

<string>{"security.osclientcerts.autoload": true,"security.default_personal_cert": "Select Automatically"}</string>

**Comment by Mgeige1** [comment]
Michael,

I tried the following with the same results that both will not apply

 <key>Preferences</key>
 
<string>{"security.osclientcerts.autoload": true", security.default_personal_cert": "Select Automatically"}
</string>

 <key>Preferences</key>
 
<string>{"security.osclientcerts.autoload": true," security.default_personal_cert": "Select Automatically"}
</string>

**Comment by mkaply** [comment]
That's odd. I just tried this configuration on my Mac and it was applying for me. my plist file is:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>EnterprisePoliciesEnabled</key>
  <true/>
  <key>Preferences</key>
  <string>{"security.osclientcerts.autoload": true,"security.default_personal_cert": "Select Automatically"}</string>
</dict>
</plist>
```

**Comment by mkaply** [comment]
I see some quote isses in your post above. Missing quote before security in the first one, and a space before security in the second one.

**Comment by Mgeige1** [comment]
I see that now, let me retest

**Comment by Mgeige1** [comment]
OK that works,  I appreciate it

**Comment by olphger** [comment]
Its fixed in 78.3.1 ESR - Thank you very much
